syntax = "proto3";

package products;

service Products{

rpc GetRowsByColumn(SelectOneMessage) returns (generalResponse){}
rpc GetRowByMultiColumns(SelectManyMessage) returns (generalResponse){}
rpc InsertProduct(InsertMessage) returns (generalResponse){}
rpc UpdateRowByColumn(UpdateMessage) returns (generalResponse){}
rpc UpdateRowByMulti(UpdateManyMessage) returns (generalResponse){}
rpc DeleteRow(DeleteMessage) returns (generalResponse){}

rpc SearchProduct(SearchProductsRequestMessage) returns (SearchProductsResponseMessage){}

}


message SelectOneMessage{
    string table_name = 1;
    string column = 2;
    string search_value = 3;
    string selected_columns = 4;
}

message SelectManyMessage{
    string table_name = 1;
    string columns = 2;
    string search_values = 3;
    int32 return_index = 4;
}

message InsertMessage{
    string table_name = 1;
    string columns = 2;
    string values = 3;
}
message UpdateMessage{
    string table_name = 1;
    string columns = 2;
    string values = 3;
    string condition_col = 4;
    int32 condition_val = 5;
}

message UpdateManyMessage{
    string table_name = 1;
    string columns = 2;
    string values = 3;
    string condition_cols = 4;
    string condition_vals = 5;
}
message DeleteMessage{
    string table_name = 1;
    string condition_cols = 4;
    string condition_vals = 5;
}

message generalResponse{
    string msg=1;
}

message SearchProductsRequestMessage {
  string item_category = 1;
  repeated string keywords = 2;
}

message Product {
  int32 id = 1;
  string item_name = 2;
  int32 seller_id = 3;
  int32 item_category = 4;
  repeated string keywords = 5;
  string condition = 6;
  string sale_price = 7;
  int32 quantity = 8;
  int32 thumbs_up_count=9;
  int32 thumbs_down_count=10;
}

message SearchProductsResponseMessage {
  repeated Product products = 1;
}